### DISCUSSION TITLE
How to create a custom logout flow?

### DISCUSSION THREAD

**Discussion by Nickk4** [keycloak/keycloak#22407] [Q&A] [answered]
When a user clicks log out in our application, we direct to `http://{domain-name}/auth/realms/{realm-name}/protocol/openid-connect/logout?redirect_uri=encodedRedirectUri `. KeyCloak then asks for confirmation if the user really wants to log out. How can I disable this confirmation?

I assumed this is to be done by adapting the 'browser' Authentication flow. However, I can't find anything about logging out in any of the authentication flows.

I'm on version 21.0.1. My current browser flow looks as follows:

![image](https://github.com/keycloak/keycloak/assets/10728166/d8eb2108-5641-4bae-a17b-1a4c42fb001c)

**Comment by justin-tay** [comment]
I think if you want to skip the logout confirmation page the way to do it is to pass the `id_token_hint` to the logout endpoint.

There's some more detailed information in the following discussion

- https://github.com/keycloak/keycloak/discussions/12183

The codes with the actual logic is here

https://github.com/keycloak/keycloak/blob/baac060eb1036f0702309854de9896f11ea3cfeb/services/src/main/java/org/keycloak/protocol/oidc/endpoints/LogoutEndpoint.java#L163

The following is the relevant documentation

https://www.keycloak.org/docs/latest/server_admin/#rp-initiated-logout

The documentation has this note 

> The user might be optionally requested to confirm the logout in case the `id_token_hint` parameter was not used

**Comment by Nickk4** [comment]
Thanks, I think I've implemented what is suggested there. I'm using the following log out url: `https://auth.mydomain.com/realms/testing/protocol/openid-connect/logout?login_hint=myusername&post_logout_redirect_uri=http://localhost:3001?loggedout=true&client_id=account`.

But KeyCloak still asks for confirmation upon logging out.

**Reply by danielFesenmeyer** [reply]
It seems like your logout URL does not contain id_token_hint, then logout won't work without confirmation.

**Reply by Nickk4** [reply]
Sorry, that was a typo. I'm using: `https://auth.mydomain.com/realms/testing/protocol/openid-connect/logout?id_token_hint=myusername&post_logout_redirect_uri=http://localhost:3001?loggedout=true&client_id=account`. But this returns "Invalid parameter: id_token_hint".

**Reply by danielFesenmeyer** [reply, marked as answer]
Ok, then it seems that the `id_token_hint` is wrong. It must a valid ID token for the client used for logout (valid, with the exception that it may be expired). The typical way to get that ID token is to extract it from the token endpoint response (field "id_token"). For more details, see https://www.keycloak.org/docs/latest/securing_apps/#token-endpoint
